src/v/transform/logging/log_manager.cc

```diff
@@ -0,0 +1,297 @@
+            if (!buffer_units) {
+                buffer_units.emplace(std::move(e.units));
+            } else {
+                buffer_units->adopt(std::move(e.units));
+            }
+
+            // reactor will stall if we try to serialize a whole lot of
+            // messages all at once
+            co_await ss::maybe_yield();
+        }
+
+        co_return io::json_batch{
+          model::transform_name{name().data(), name().size()},
+          std::move(ev_json),
+          std::move(*buffer_units)};
+    }
+
+    ss::future<> do_flush(model::partition_id pid, io::json_batches events) {
+        // NOTE(oren): At this point, semaphore units for occupied buffer
+        // capacity have been adopted by `json_batch`es. These units will be
+        // released (i.e. buffer capacity freed) only once those records have
+        // been produced.
+        co_await _client->write(pid, std::move(events));
+        co_return;
+    }
+
+private:
+    ss::abort_source* _as = nullptr;
+    transform::logging::client* _client = nullptr;
+    config::binding<std::chrono::milliseconds> _interval_ms;
+    simple_time_jitter<ClockType> _jitter;
+    ss::condition_variable _wakeup_signal{};
+};
+
+template class flusher<ss::lowres_clock>;
+template class flusher<ss::manual_clock>;
+
+} // namespace detail
+
+template<typename ClockType>
+manager<ClockType>::manager(
+  model::node_id self,
+  std::unique_ptr<client> c,
+  size_t bc,
+  config::binding<size_t> ll,
+  config::binding<std::chrono::milliseconds> fi)
+  : _self(self)
+  , _client(std::move(c))
+  , _line_limit_bytes(std::move(ll))
+  , _buffer_limit_bytes(bc)
+  , _buffer_low_water_mark(_buffer_limit_bytes / lwm_denom)
+  , _buffer_sem(_buffer_limit_bytes, "Log manager buffer semaphore")
+  , _flush_interval_ms(std::move(fi)) {}
+
+template<typename ClockType>
+manager<ClockType>::~manager() = default;
+
+template<typename ClockType>
+ss::future<> manager<ClockType>::start() {
+    _flusher = std::make_unique<detail::flusher<ClockType>>(
+      &_as, _client.get(), _flush_interval_ms);
+    return _flusher->template start<>(_gate, &_log_buffers);
+}
+
+template<typename ClockType>
+ss::future<> manager<ClockType>::stop() {
+    _as.request_abort();
+    co_await _flusher->stop();
+    if (!_gate.is_closed()) {
+        co_await _gate.close();
+    }
+}
+
+template<typename ClockType>
+bool manager<ClockType>::check_lwm() const {
+    return _buffer_sem.available_units() <= _buffer_low_water_mark;
+}
+
+template<typename ClockType>
+void manager<ClockType>::enqueue_log(
+  ss::log_level level,
+  model::transform_name_view transform_name,
+  std::string_view message) {
+    auto msg_len =
+      [lim = _line_limit_bytes()](std::string_view message) -> size_t {
+        return std::min(lim, message.size());
+    };
+
+    auto validate_msg =
+      [&msg_len](std::string_view message) -> std::optional<ss::sstring> {
+        auto sub_view = message.substr(0, msg_len(message));
```

Comment:
Technically this can leave invalid UTF-8 due to a multiple byte character being cut in the middle, but personally I don't think we need to worry about it.